src/transform/src/fusion/reduce.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Fuses reduce operators with parent operators if possible.
+use crate::TransformArgs;
+use expr::{MirRelationExpr, MirScalarExpr};
+
+/// Fuses reduce operators with parent operators if possible.
+#[derive(Debug)]
+pub struct Reduce;
+
+impl crate::Transform for Reduce {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl Reduce {
+    /// Fuses reduce operators with parent operators if possible.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic: _,
+            expected_group_size: _,
+        } = relation
+        {
+            if let MirRelationExpr::Reduce {
+                input: inner_input,
+                group_key: inner_group_key,
+                aggregates: inner_aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = &mut **input
+            {
+                // Collect all columns referenced by outer
+                let mut outer_cols = vec![];
+                for expr in group_key.iter() {
+                    expr.visit(&mut |e| {
+                        if let MirScalarExpr::Column(i) = e {
+                            outer_cols.push(*i);
+                        }
+                    });
+                }
+
+                // We can fuse reduce operators as long as the outer one doesn't
+                // group by an aggregation performed by the inner one.
+                if outer_cols.iter().any(|c| *c >= inner_group_key.len()) {
+                    return;
+                }
+
+                if aggregates.is_empty() && inner_aggregates.is_empty() {
```

Comment:
In a followup patch we could introduce a `translate(AggregateExpr, inner_key, outer_key) -> Option<AggregateExpr>` function that given an aggregate in the outer reduction, returns an equivalent aggregate after the fusion if possible. Then in order for two Reduce operators to be all the aggregates from the outer one must return a valid translated aggregate.

Some cases are trivial like MIN and MAX as long as they only reference columns from the inner key. COUNT and SUM become COUNT(DISTINCT) and SUM(DISTINCT) respectively, again, as long as they only reference columns from the inner key, not aggregated columns of the inner reduction.